OemPkg/OemConfigPolicyCreatorPei/OemConfigPolicyCreatorPei.c

```diff
@@ -0,0 +1,246 @@
+    }
+
+    // Validate the value from flash meets the constraints of the knob
+    if ((Status == EFI_SUCCESS) && (gKnobData[i].Validator != NULL)) {
+      if (!gKnobData[i].Validator (gKnobData[i].CacheValueAddress)) {
+        // If it doesn't, we will set the value to the default value
+        DEBUG ((DEBUG_ERROR, "Config knob %a failed validation!\n", gKnobData[i].Name));
+        CopyMem (gKnobData[i].CacheValueAddress, gKnobData[i].DefaultValueAddress, gKnobData[i].ValueSize);
+      }
+    }
+
+    ConfListPtr = ((UINT8 *)*ConfPolicy) + Offset;
+
+    VarListSize = NeededSize - Offset;
+    Status      = ConvertVariableEntryToVariableList (&VarListEntry, ConfListPtr, &VarListSize);
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a failed to convert variable entry to var list! - %r\n", __FUNCTION__, Status));
+      ASSERT (FALSE);
+      goto CreatePolicyExit;
+    }
+
+    Offset += VarListSize;
+  }
+
+  if (Offset != NeededSize) {
+    // oops we messed up the math, may have corrupted memory...
+    DEBUG ((DEBUG_ERROR, "%a expected ConfPolicy size %x does not match actual size %x!\n", __FUNCTION__, NeededSize, Offset));
+    ASSERT (Offset == NeededSize);
+    Status = EFI_ABORTED;
+    goto CreatePolicyExit;
+  }
+
+CreatePolicyExit:
+  if (EFI_ERROR (Status)) {
+    *ConfPolicySize = 0;
+
+    if (*ConfPolicy != NULL) {
+      FreePool (*ConfPolicy);
+    }
+  }
+
+  return Status;
+}
+
+/**
+  Module entry point that will check configuration data and publish them to policy database.
+
+  @param FileHandle                     The image handle.
+  @param PeiServices                    The PEI services table.
+
+  @retval Status                        From internal routine or boot object, should not fail
+**/
+EFI_STATUS
+EFIAPI
+OemConfigPolicyCreatorPeiEntry (
+  IN EFI_PEI_FILE_HANDLE     FileHandle,
+  IN CONST EFI_PEI_SERVICES  **PeiServices
+  )
+{
+  EFI_STATUS  Status;
+  POLICY_PPI  *PolPpi        = NULL;
+  VOID        *ConfPolicy    = NULL;
+  UINT16      ConfPolicySize = 0;
+
+  DEBUG ((DEBUG_INFO, "%a - Entry.\n", __FUNCTION__));
+
+  // First locate policy ppi.
+  Status = PeiServicesLocatePpi (&gPeiPolicyPpiGuid, 0, NULL, (VOID *)&PolPpi);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to locate Policy PPI - %r\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    goto Exit;
+  }
+
+  // Oem can choose to do any Oem specific things to config here such as enforcing static only config or
+  // selecting a configuration profile based on some criteria
+
+  Status = CreateConfPolicy (&ConfPolicy, &ConfPolicySize);
+
+  if (EFI_ERROR (Status) || (ConfPolicy == NULL) || (ConfPolicySize == 0)) {
+    DEBUG ((DEBUG_ERROR, "%a CreateConfPolicy failed! Status (%r)\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    goto Exit;
+  }
+
+  // Publish immutable config policy
+  // Policy Service will publish the gOemConfigPolicyGuid so that the Silicon Policy Creator can consume our

```

Comment:
Comment outdated?

Reply:
This is correct, but maybe confusingly worded. I can reword. The intent is to specify that PolicyService will receive gOemConfigPolicyGuid and will publish it as a PPI so that other modules can take a Depex on it.